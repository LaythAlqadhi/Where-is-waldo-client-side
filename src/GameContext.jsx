import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';

const initialState = {
  state: { loading: false, error: null },
  status: { isStarted: false, isEnded: true },
  message: null,
};

export const GameContext = createContext(initialState);

export function GameProvider({ children }) {
  const [state, setState] = useState({ loading: false, error: null });
  const [status, setStatus] = useState({
    isStarted: false,
    isEnded: false,
    message: null,
  });
  const [message, setMessage] = useState(null);

  return (
    <GameContext.Provider
      value={{
        state,
        setState,
        status,
        setStatus,
        message,
        setMessage,
      }}
    >
      {children}
    </GameContext.Provider>
  );
}

GameProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
