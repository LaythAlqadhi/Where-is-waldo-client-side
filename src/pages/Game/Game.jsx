import React, { useState, useEffect, useContext } from 'react';
import Image from '../../components/Image/Image';
import Modal from '../../components/Modal/Modal';
import CharacterSelection from '../../components/CharacterSelection/CharacterSelection';
import { GameContext } from '../../GameContext';
import style from './Game.module.css';

export default function Game() {
  const { state, setState, status } = useContext(GameContext);
  const [position, setPosition] = useState({ x: null, y: null });
  const [focus, setFocus] = useState(false);
  const [id, setId] = useState(null);
  const [characters, setCharacters] = useState(null);

  useEffect(() => {
    const fetchData = async () => {
      const API_URL = 'https://where-is-waldo.adaptable.app/v1';

      try {
        setState({ ...state, loading: true });

        const response = await fetch(`${API_URL}/start`, { mode: 'cors' });

        if (response.status >= 400) {
          throw new Error('Server error');
        }

        const result = await response.json();

        setId(result.id);
        setCharacters(result.characters);
      } catch (err) {
        setState({ ...state, error: err.message });
      } finally {
        setState({ ...state, loading: false });
      }
    };

    if (status.isStarted) fetchData();
  }, [status.isStarted]);

  return (
    <div className={style.container}>
      <Image setPosition={setPosition} focus={focus} setFocus={setFocus} />
      {status.isEnded && <Modal id={id} />}
      <CharacterSelection
        id={id}
        focus={focus}
        setFocus={setFocus}
        position={position}
        characters={characters}
        setCharacters={setCharacters}
      />
    </div>
  );
}
