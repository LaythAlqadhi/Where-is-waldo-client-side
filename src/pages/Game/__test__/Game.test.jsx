import React from 'react';
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import { GameContext } from '../../../GameContext';
import Game from '../Game';

const mockData = {
  id: 'mockId',
  characters: {
    name: 'mockCharacter',
    coordinates: {
      x: 50,
      y: 50,
    },
  },
};

beforeEach(() => {
  global.fetch = vi.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve(mockData),
      status: 200,
    }),
  );
});

it('renders Game component correctly with mock data', () => {
  const setState = vi.fn();

  render(
    <GameContext.Provider
      value={{
        state: { loading: false, error: null },
        setState,
        status: { isStarted: true, isEnded: false },
      }}
    >
      <Game />
    </GameContext.Provider>,
  );

  expect(screen.container).toMatchSnapshot();
});
