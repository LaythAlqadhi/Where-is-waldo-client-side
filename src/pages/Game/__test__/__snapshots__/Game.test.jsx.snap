// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Game component correctly with mock data 1`] = `undefined`;
